workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/WorkflowUiViewState.kt

```diff
@@ -0,0 +1,192 @@
+ *
+ * @throws IllegalStateException if [bindShowRendering] has not been called.
+ */
+@WorkflowUiExperimentalApi
+public fun <RenderingT : Any> View.showRendering(
+  rendering: RenderingT,
+  viewEnvironment: ViewEnvironment
+) {
+  workflowTagAsStarted.let { tag ->
+    check(tag.showing.matches(rendering)) {
+      "Expected $this to be able to show rendering $rendering, but that did not match " +
+        "previous rendering ${tag.showing}. " +
+        "Consider using WorkflowViewStub to display arbitrary types."
+    }
+
+    // Update the tag's rendering and viewEnvironment.
+    workflowTag = Started(rendering, viewEnvironment, tag.showRendering)
+    // And do the actual showRendering work.
+    tag.showRendering.invoke(rendering, viewEnvironment)
+  }
+}
+
+/**
+ * Returns the most recent rendering shown by this view cast to [RenderingT],
+ * or null if [bindShowRendering] has never been called.
+ *
+ * @throws ClassCastException if the current rendering is not of type [RenderingT]
+ */
+@WorkflowUiExperimentalApi
+public inline fun <reified RenderingT : Any> View.getRendering(): RenderingT? {
+  // Can't use a val because of the parameter type.
+  return when (val showing = workflowTagOrNull?.showing) {
+    null -> null
+    else -> showing as RenderingT
+  }
+}
+
+/**
+ * Returns the most recent [ViewEnvironment] applied to this view, or null if [bindShowRendering]
+ * has never been called.
+ */
+@WorkflowUiExperimentalApi
+public val View.environment: ViewEnvironment?
+  get() = workflowTagOrNull?.environment
+
+/**
+ * Returns the function set by the most recent call to [bindShowRendering], or null
+ * if that method has never been called.
+ */
+@WorkflowUiExperimentalApi
+public fun <RenderingT : Any> View.getShowRendering(): ViewShowRendering<RenderingT>? {
+  return workflowTagOrNull?.showRendering
+}
+
+@WorkflowUiExperimentalApi
+internal var View.starter: (View) -> Unit
+  get() = workflowTagAsNew.starter
+  set(value) {
+    workflowTag = workflowTagAsNew.copy(starter = value)
+  }
+
+@WorkflowUiExperimentalApi
+@PublishedApi
+internal val View.workflowTagOrNull: WorkflowUiViewState<*>?
+  get() = getTag(R.id.workflow_ui_view_state) as? WorkflowUiViewState<*>
+
+@WorkflowUiExperimentalApi
+private var View.workflowTag: WorkflowUiViewState<*>
+  get() = workflowTagOrNull ?: error(
+    "Expected $this to have been built by a ViewFactory. " +
+      "Perhaps the factory did not call View.bindShowRendering."
+  )
+  set(value) = setTag(R.id.workflow_ui_view_state, value)
+
+@WorkflowUiExperimentalApi
+private val View.workflowTagAsNew: New<*>
+  get() = workflowTag as? New<*> ?: error(
+    "Expected $this to be un-started, but View.start() has been called"
+  )
+
+@WorkflowUiExperimentalApi
+private val View.workflowTagAsStarted: Started<*>
+  get() = workflowTag as? Started<*> ?: error(
+    "Expected $this to have been started, but View.start() has not been called"
+  )
```

Comment:
Goodness me this is the first time I have deeply pined for Swift's protocol extensions.

I get why we need all of this as extensions on the View but it is _very_ difficult to follow the logic with all these (albeit handy) extension shortcuts.

I guess it's not possible (given the coupling) to extend `View` to a `WorkflowManagedView` to have all these extensions in a more understandable form?

Reply:
Nope, this approach is entirely about avoiding the need for custom View classes.

We _could_ create a facade interface that just wrapped a plain old View. I don't think it's worth the effort. Feature developers basically never bump into this stuff, thanks to `WorkflowViewStub` and such. That will be even more true when the UI overhaul is done, I think.

But just to scratch the nerdsnipe itch, here's a sketch of what that could be.

```kotlin
fun <RenderingT> ViewRegistry.buildView(...): WorkflowView<RenderingT>

interface WorkflowView<RenderingT> {
  fun asView(): View
  
  fun start()

  val rendering: RenderingT
  val environment: ViewEnvironment

  fun update(
    rendering: RenderingT,
    environment: ViewEnvironment
  )
  // ...
}

fun <RenderingT> View.asWorkflowView(): WorkflowView<T>? {
  // Looks for the tag and confirms the type is right? 
}
```

Still thinking this would be a distraction, but I'm wavering. Dammit Steve, are you proud that you've done this to me?

Reply:
And `WorkflowViewStub` could implement `WorkflowView<*>`! All the stock containers could!

Reply:
Sigh, tracked: #606

Reply:
> Goodness me this is the first time I have deeply pined for Swift's protocol extensions.

Man, I've been wishing for those since I left Objective-C for Java.

Reply:
Haha. Amazing. Very proud. I get your point though that the audience is small. This _does_ make it a lot easier to grok though IMHO.